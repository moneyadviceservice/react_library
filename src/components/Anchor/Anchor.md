Renders an Anchor element.

```jsx
import { Anchor } from '@moneypensionservice/directories';

<Anchor 
  href="https://www.google.co.uk"
  target="_blank"
>
  I'm a link to Google!
</Anchor>
```
