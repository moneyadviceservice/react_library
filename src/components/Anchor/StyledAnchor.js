import styled from 'styled-components'
import { Paragraph } from '../Paragraph'

const StyledAnchor = styled(Paragraph)`
  color: ${props => props.colorProp || props.theme.colors.linkDefault};
  text-decoration: none;

  &:hover,
  &:focus {
    text-decoration: underline;
  }

  &:hover {
    outline: none;
  }

  &:focus {
    outline: solid 0.1875rem ${props => props.theme.colors.focusBg};
    color: ${props => props.colorProp || props.theme.colors.focusBg};
  }

  // Removes active link colour that is applied in Mobile Safari and adds kerning
  &[href^='tel'] {
    color: inherit;
    letter-spacing: -1px;
    text-decoration: none;
  }
`

export default StyledAnchor
