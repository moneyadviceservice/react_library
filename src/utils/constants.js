export const breakpoints = {
  xs: 1,
  sm: 48, // 768px
  md: 64, // 1024px
  lg: 90, // 1440px
  xl: 120, // 1920px
}

export const dimensions = ['xs', 'sm', 'md', 'lg', 'xl']

export const mediaQuery = 'only screen'

export const ALIGN_SELF_MAP = {
  center: 'center',
  end: 'flex-end',
  start: 'flex-start',
  stretch: 'stretch',
}

export const SPACING_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
]

