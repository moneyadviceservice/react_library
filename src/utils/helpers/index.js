export * from './background'
export * from './responsive'
export * from './flexbox'
export * from './styles'
export * from './mediaQueries'
export * from './sizing'
export * from './refs'
