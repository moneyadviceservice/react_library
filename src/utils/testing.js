import { render, act } from '@testing-library/react'
import GlobalTheme from '../components/ThemeProvider'

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalTheme, ...options })

const waitRender = async (ui, options) => {
  const { container, ...rest } = customRender(ui, options)

  await act(async () => {
    await Promise.resolve(container)
  })

  return { container, ...rest }
}

export * from '@testing-library/react'

export { customRender as render, waitRender }
