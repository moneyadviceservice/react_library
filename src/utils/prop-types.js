import PropTypes from 'prop-types'
import { SPACING_SIZES } from './constants'

/** spacing prop.type */
const spacingProp = PropTypes.shape({
  bottom: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  horizontal: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  left: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  right: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  top: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  vertical: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
})

export const genericPropTypes = {
  a11yTitle: PropTypes.string,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  margin: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  padding: PropTypes.oneOfType([
    PropTypes.oneOf([...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  border: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
}

export function genericPropsDefaults({
  a11yTitle = null,
  alignSelf = null,
  margin = null,
  padding = null,
  border = null,
} = {}) {
  return {
    a11yTitle: a11yTitle,
    alignSelf: alignSelf,
    margin: margin,
    padding: padding,
    border: border,
  }
}
