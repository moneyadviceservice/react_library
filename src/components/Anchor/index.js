import React from 'react'
import PropTypes from 'prop-types'
import { genericPropTypes, genericPropsDefaults } from '../../utils/prop-types'
import StyledAnchor from './StyledAnchor'

const Anchor = ({ a11yTitle, as, color, weight, width, ...rest }) => {
  return (
    <StyledAnchor
      aria-label={a11yTitle}
      as={as}
      colorProp={color}
      weightProp={weight}
      widthProp={width}
      {...rest}
    />
  )
}

// Documentation
Anchor.propTypes = {
  /** The DOM tag or react component to use for the element. */
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func]),
  /** Content inside component. */
  children: PropTypes.node,
  /** Changes the text color. */
  color: PropTypes.string,
  /** Set text style to italic. */
  italic: PropTypes.bool,
  /** Sets the line-height property. */
  lineHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  /** Sets the horizontal alignment. Values: start, end, left, right, center, justify, justify-all, or match-parent. */
  textAlign: PropTypes.oneOfType([
    PropTypes.oneOf([
      'start',
      'end',
      'left',
      'right',
      'center',
      'justify',
      'justify-all',
      'match-parent',
    ]),
    PropTypes.object,
  ]),
  /** Strike through the text. */
  strike: PropTypes.bool,
  /** Sets the font-size property. */
  textSize: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  /** Underlines the text. */
  underline: PropTypes.bool,
  /** Sets the font-weight property. */
  weight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** Sets custom width to the element. */
  width: PropTypes.string,
  ...genericPropTypes,
}

Anchor.defaultProps = {
  as: 'a',
  color: null,
  children: null,
  italic: false,
  lineHeight: null,
  strike: false,
  textAlign: null,
  textSize: null,
  underline: false,
  weight: null,
  width: '100%',
  ...genericPropsDefaults(),
}

/** @component */
export { Anchor }
