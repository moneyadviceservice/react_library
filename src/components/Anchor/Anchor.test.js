import React from 'react'
import { render } from '../../utils/testing'
import { Anchor } from '../Anchor'

describe('<Anchor />', () => {
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      const { container } = render(<Anchor>example text</Anchor>)
      expect(container).toMatchSnapshot()
    })
    it('should match snapshot with italic and textSize props', () => {
      const { container } = render(
        <Anchor italic textSize="12px">
          example text
        </Anchor>
      )
      expect(container).toMatchSnapshot()
    })
    it('should match snapshot with weight and underline props', () => {
      const { container } = render(
        <Anchor weight="700" underline>
          example text
        </Anchor>
      )
      expect(container).toMatchSnapshot()
    })
  })

  describe('href prop', () => {
    it('should apply href property to the anchor', () => {
      const { container, getByText } = render(
        <Anchor href="https://www.test.com/">example</Anchor>
      )
      // snapshot with href
      expect(container).toMatchSnapshot()
      // href
      expect(getByText('example').href).toBe('https://www.test.com/')
    })
  })
})
