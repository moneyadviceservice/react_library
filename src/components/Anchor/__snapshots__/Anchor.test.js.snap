// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Anchor /> Snapshots should match snapshot 1`] = `
<div>
  <a
    class="StyledParagraph-sc-1toaej-0 gpltyO StyledAnchor-sc-1qj1ci5-0 fNhjgf"
  >
    example text
  </a>
</div>
`;

exports[`<Anchor /> Snapshots should match snapshot with italic and textSize props 1`] = `
<div>
  <a
    class="StyledParagraph-sc-1toaej-0 blbiyd StyledAnchor-sc-1qj1ci5-0 fNhjgf"
  >
    example text
  </a>
</div>
`;

exports[`<Anchor /> Snapshots should match snapshot with weight and underline props 1`] = `
<div>
  <a
    class="StyledParagraph-sc-1toaej-0 koRzUD StyledAnchor-sc-1qj1ci5-0 fNhjgf"
  >
    example text
  </a>
</div>
`;

exports[`<Anchor /> href prop should apply href property to the anchor 1`] = `
<div>
  <a
    class="StyledParagraph-sc-1toaej-0 gpltyO StyledAnchor-sc-1qj1ci5-0 fNhjgf"
    href="https://www.test.com/"
  >
    example
  </a>
</div>
`;
